### PR TITLE
Add Remote Reward Server Feature 

### LINKED ISSUE #269: Support reward model on a remote server
Suppose we have a scenario where we have deployed a reward model (e.g, [ArmoRM-Llama3-8B-v0.1](https://huggingface.co/RLHFlow/ArmoRM-Llama3-8B-v0.1), GPT-4 as a judge) on a remote server. We want to call these reward model services in RLHF. [OpenRLHF](https://openrlhf.readthedocs.io/en/latest/remote_rm.html) provides an HTTP interface to achieve this. Do we have plans to support a similar feature?

### PR DESCRIPTION
**Pull request by YuchenFan48**
# Description

This PR introduces support for remote generative models to enhance the verification and reward assignment process in VERL. Previously, VERL only supported rule-based rewards for verification. With this update, we enable more flexible and dynamic reward mechanisms using generative models.

# Key Changes

## Bug Fix:
- Resolved an issue in `main_ppo.py` where the `val_dataloader` was incorrectly processed in batches. This ensures proper handling of validation data during training.

## New Features:
- Added a **Generative Reward Manager** in `workers/reward_manager` to handle reward generation using remote generative models.
- Implemented a corresponding `compute_score` function in `utils/reward_score` to calculate scores based on generative model outputs.

## Issue Resolution:
- Closes [#269](https://github.com/volcengine/verl/issues/269) and [#229](https://github.com/volcengine/verl/issues/229): Adds support for remote generative reward mechanisms.

### CODE REVIEW

**Comment by YuchenFan48** [comment]
Hi, do you find this pr useful for verl?

**Comment by vermouth1992** [comment]
Please resolve conflict first

**Comment by YuchenFan48** [comment]
Conflicts reloved. Thanks.

File: verl/trainer/ppo/ray_trainer.py [hunk 1]
```diff
@@ -570,15 +570,11 @@ def _create_dataloader(self):
...
-        self.val_dataloader = StatefulDataLoader(
-            dataset=self.val_dataset,
-            # Validation datasets are sent to inference engines as a whole batch,
-            # which will schedule the memory themselves.
-            batch_size=len(self.val_dataset),
-            num_workers=8,
-            shuffle=False,
-            drop_last=False,
-            collate_fn=collate_fn)
+        self.val_dataloader = DataLoader(dataset=self.val_dataset,
```
Comment (vermouth1992) [inline verl/trainer/ppo/ray_trainer.py]: I guess we should use StatefulDataLoader instead of normal Pytorch DataLoader to support save/load dataloader state.

File: verl/trainer/ppo/ray_trainer.py [hunk 2]
```diff
@@ -636,7 +632,9 @@ def _validate(self):
 
         for test_data in self.val_dataloader:
             test_batch = DataProto.from_single_dict(test_data)
-
+            # print example of the first batch
```
Comment (vermouth1992) [inline verl/trainer/ppo/ray_trainer.py]: Could you remove this because we already log the cases to wandb

File: verl/utils/prompts.txt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2024 Bytedance Ltd. and/or its affiliates
```
Comment (vermouth1992) [inline verl/utils/prompts.txt:1]: Why do we need this file?

Comment (YuchenFan48) [reply verl/utils/prompts.txt:1]: It is the file where the user puts the scoring prompt.

File: verl/trainer/ppo/ray_trainer.py (hunk 1 above)
Comment (YuchenFan48) [reply verl/trainer/ppo/ray_trainer.py]: I have replaced DataLoader with StatefulDataLoader.

File: verl/trainer/ppo/ray_trainer.py (hunk 2 above)
Comment (YuchenFan48) [reply verl/trainer/ppo/ray_trainer.py]: Codes to print examples are disabled.

File: verl/trainer/ppo/ray_trainer.py [hunk 4]
```diff
@@ -570,15 +570,11 @@ def _create_dataloader(self):
...
-        self.val_dataloader = StatefulDataLoader(
-            dataset=self.val_dataset,
-            # Validation datasets are sent to inference engines as a whole batch,
-            # which will schedule the memory themselves.
-            batch_size=len(self.val_dataset),
-            num_workers=8,
-            shuffle=False,
-            drop_last=False,
-            collate_fn=collate_fn)
+        self.val_dataloader = StatefulDataLoader(dataset=self.val_dataset,
```
Comment (vermouth1992) [inline verl/trainer/ppo/ray_trainer.py]: Could you revert this to what main has?

Comment (vermouth1992) [reply verl/trainer/ppo/ray_trainer.py]: Also need to perform format

File: examples/ppo_trainer/run_ppo.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
+set -x
```
Comment (vermouth1992) [inline examples/ppo_trainer/run_ppo.sh]: What's the purpose of this file?

File: verl/utils/prompts.txt (hunk 3 above)
Comment (vermouth1992) [reply verl/utils/prompts.txt:1]: Can this scoring prompt be passes in other ways? e.g, from commandline?

File: examples/ppo_trainer/run_ppo.sh (hunk 5 above)
Comment (YuchenFan48) [reply examples/ppo_trainer/run_ppo.sh]: Bash file removed

File: verl/trainer/ppo/ray_trainer.py (hunk 4 above)
Comment (YuchenFan48) [reply verl/trainer/ppo/ray_trainer.py]: Codes revert to main branch and use standard format.

File: verl/utils/prompts.txt (hunk 3 above)
Comment (YuchenFan48) [reply verl/utils/prompts.txt:1]: Sure. You can pass scoring prompt from command line using reward_model.reward_manager_config.scoring_prompt

**Comment by vermouth1992** [comment]
LGTM now. There are still conflicts though. Make sure to run format script and I will tigger the workflow. Thanks.

**Comment by YuchenFan48** [comment]
Format scripts used. Thanks.

**Comment by vermouth1992** [comment]
Could you resolve the conflict?

**Comment by YuchenFan48** [comment]
> Could you resolve the conflict?

If we use generative rewards, we should pass more arguments than rule-based rewards. So passing a config to a generative reward manager may be plausible.

**Comment by YuchenFan48** [comment]
Conflicts resolved. I added a configuration file for generative rewards.

**Comment by vermouth1992** [comment]
Could you provide an example that uses GenRM to train a math task that can improve the scores? It would be great help

**Comment by YuchenFan48** [comment]
In my opinion, strong LLMs can serve as a math interpreter as current rule-based math interpreter do. So with a strong LLM, or LRM, it should be at least as good as rule-based rewards. More importantly, LLMs or LRMs,e.g., RM-R1, JudgeLRM, or LLM-as-a-Judge, can help training open-domain abilities. However, I don't carry experiments on this. Sorry about this.

**Comment by YuchenFan48** [comment]
Hey. How is the review going

File: verl/workers/reward_manager/generative.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def _compute_score(data_source, solution_str, ground_truth,
+                   extra_info, config) -> torch.Tensor:
+    loop = asyncio.get_event_loop()
+    return loop.run_until_complete(process_data_async(data_source, solution_str, ground_truth, extra_info, config))
+
+
+class GenerativeRewardManager:
+
```
Comment (eric-haibin-lin) [inline verl/workers/reward_manager/generative.py]: please add docstring to this new class

Comment (eric-haibin-lin) [reply verl/workers/reward_manager/generative.py]: could you add a unit test for this class, construct a manager, and invoke it?
additionally if you have end-to-end example that shows the implementation works for a RL dataset it would be great

Comment (YuchenFan48) [reply verl/workers/reward_manager/generative.py]: Hi. Doc String is added for the class. And I implemented a unit test for generative rewards in verl/tests/reward_score/test_gen_rewards.py. And a script is added in verl/examples/grpo_trainer/run_qwen2-7b_gen_reward.sh. Before running the bash, fill in the attributes in verl/verl/trainer/config/generative_reward_config.yaml, or we can merge it into ppo_trainer.yaml and pass a config to the reward manager. Which one is better?

File: verl/trainer/config/generative_reward_config.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+api_key: null
```
Comment (vermouth1992) [inline verl/trainer/config/generative_reward_config.yaml:1]: Could you provide an example to use local deployed model and add it to the CI? Thanks!

Comment (YuchenFan48) [reply verl/trainer/config/generative_reward_config.yaml:1]: Yeah. I am glad to implement local GenRM when I am free. However, I think server is the most convenient way.

Comment (YuchenFan48) [reply verl/trainer/config/generative_reward_config.yaml:1]: Maybe you can merge this version and later I will implement local GenRM

File: tests/reward_score/test_gen_rewards.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+import asyncio
+
+class TestGenerativeRewardManager(unittest.TestCase):
+    @patch('yaml.safe_load')
+    @patch('builtins.open')
+    def setUp(self, mock_open, mock_yaml_load):
+        # Import the necessary classes
+        from transformers import PreTrainedTokenizer
+        import sys
+        sys.path.append("/fs-computility/mabasic/fanyuchen/verl")
```
Comment (eric-haibin-lin) [inline tests/reward_score/test_gen_rewards.py]: i think this should be removed, otherwise CI would fail

Comment (YuchenFan48) [reply tests/reward_score/test_gen_rewards.py]: I removed the sys content. You can try merging again.

**Comment by YuchenFan48** [comment]
Hi. Why does the merge fail several times? Could you provide some suggestions?

**Comment by haorannlp** [comment]
> Hi. Why does the merge fail several times? Could you provide some suggestions?

have tested this myself. One problem is that the script exits but the program is still running (GPU is still actively running)
```
[36m(WorkerDict pid=62541)[0m rollout and sharding manager init done sharding_manager: <verl.workers.sharding_manager.megatron_vllm.MegatronVLLMShardingManager object at 0x7fb144680090>[32m [repeated 15x across cluster][0m
[36m(TaskRunner pid=50661)[0m test_gen_batch meta info: {'eos_token_id': 151643, 'pad_token_id': 151665, 'recompute_log_prob': False, 'do_sample': False, 'validate': True}
[36m(WorkerDict pid=62539)[0m NCCL version 2.21.5+cuda12.4
[36m(TaskRunner pid=50661)[0m validation generation end
[36m(WorkerDict pid=62542)[0m NCCL version 2.21.5+cuda12.4[32m [repeated 6x across cluster][0m
Status for job 'raysubmit_j8m9mU7grA6U8y8Z': RUNNING
Status message: Job is currently running.
```

After switching to rule-based reward the program runs smoothly

File: verl/utils/reward_score/generative.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            },
+            {
+                "role": "user",
+                "content": scoring_prompt + '\n' + sequence_str
+            },
+        ]
+        # if using vllm server
+        if config['tokenizer'] and config['apply_chat_template']:
+            tokenizer = AutoTokenizer.from_pretrained(config['tokenizer'])
+            messages = tokenizer['apply_chat_template'](messages, tokenize=False, add_generation_prompt=True)
```
Comment (yzlnew) [inline verl/utils/reward_score/generative.py:54]: Why `apply_chat_template` here? Is this necessary as we are calling OpenAI compatible servers?
